### PR TITLE
crypto: Remove CRYPTO_SW_AES and CRYPTO_BLAKE2S

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary

since it's simpler to let linker remove the unused functions from the final image

## Impact

crypto

## Testing

ci

### CODE REVIEW

**Comment by acassis** [comment]
@xiaoxiang781216 I understand that your idea is to simplify the Kconfig, but I think removing these symbols will cause the compiler to spend time compiling all these files, just to the end the linker decide not to include it in the final binary.

Is there any other benefit of doing it?

File: crypto/Kconfig [hunk 1]
```diff
@@ -50,28 +50,9 @@ config CRYPTO_CRYPTODEV_HARDWARE
 	depends on CRYPTO_CRYPTODEV
 	default n
 
-config CRYPTO_SW_AES
-	bool "Software AES library"
-	depends on ALLOW_BSD_COMPONENTS
```
Comment (btashton) [inline crypto/Kconfig:55]: I generally agree here to just let the linker deal with it, but if want to actually have this `ALLOW_BSD_COMPONENTS` flag we cannot merge this in. That said I don't think this flag provides much value and I don't think it guards all BSD code.

Comment (xiaoxiang781216) [reply crypto/Kconfig:55]: Move ALLOW_BSD_COMPONENTS to config CRYPTO, since the whole crpyto driver framework come from FreeBSD.

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 I understand that your idea is to simplify the Kconfig, but I think removing these symbols will cause the compiler to spend time compiling all these files, just to the end the linker decide not to include it in the final binary.

Yes, it's a tradeoff between the build speed and user friendly. I prefer to simplify Kconfig to make the selection as automation as possible if the speed degradation is minor like this case.

> 
> Is there any other benefit of doing it?
> 

Yes, if you list nuttx/crypto, you can see that we support a dozen of soft crypto algo now. If we follow the original design principle, a dozen of option needs to be added to Kconfig, it's terrible to use them correctly for a newbie.

> Another detail is that some symbols depend on user manually selecting CONFIG_ALLOW_BSD_COMPONENTS, otherwise the final firmware cannot be 100% Apache license compatible

ALLOW_BSD_COMPONENTS should be move to "config CRYPTO" instead, since the whole crypto code base come from FreeBSD.

**Review by btashton — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@acassis the license issue is fixed, please review again.

**Review by acassis — APPROVED** [review]

**Comment by acassis** [comment]
@xiaoxiang781216 done! But I think now we have other issue: you are assuming that all Crypto API depends on BSD license, that is not true case the user are not using AES

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 done! But I think now we have other issue: you are assuming that all Crypto API depends on BSD license, that is not true case the user are not using AES

Since all other software crypto and /dev/crytpo comes from BSD, it's right to let the whole crypto subsystem depends on ALLOW_BSD_COMPONENTS.

**Comment by xiaoxiang781216** [comment]
@acassis what's your opinion.

**Comment by acassis** [comment]
@xiaoxiang781216 according with LICENSE file only AES is BSD. All NuttX files came from BSD because NuttX was BSD :-)

But they were re-licensed as Apache license. The AES is an exception because it is a foreign file. I think it is better to put CRYPTO_AES as a separated option instead forcing the users to enable BSD License to use the crypto

**Comment by acassis** [comment]
Maybe we will need to find an AES implementation under Apache License because AES is very important for many applications such as TLS

**Comment by xiaoxiang781216** [comment]
> Maybe we will need to find an AES implementation under Apache License because AES is very important for many applications such as TLS.
> 
> Here: https://github.com/Mbed-TLS/mbedtls/blob/development/library/aes.c

AES in kernel space is just used to very crypto driver framework work as design, nobody will use it in the production since:

1. Implement the real crypto AES driver if hardware support it and enable it in userspace crypto library like:
    https://github.com/apache/nuttx-apps/pull/1932
2. Or use the userspace crypto library(e.g. mbedtls, tinycypto) directly

It doesn't make sense to port a software crypto algo to kernel due to license.

**Comment by btashton** [comment]
I think the BSD flag is of zero value and should be dropped. It does not make sense.

**Comment by xiaoxiang781216** [comment]
> I think the BSD flag is of zero value and should be dropped. It does not make sense.

As I understand, how to process IP is different from BSD to Apache.

**Comment by xiaoxiang781216** [comment]
@acassis do you have more suggestion?

**Comment by xiaoxiang781216** [comment]
ping @acassis

**Comment by xiaoxiang781216** [comment]
@acassis could you merge this patch?

**Pull request merged**